The c-ares and DNSPython resolvers now raise exceptions much more
consistently with the standard resolver. Types and errnos are
substantially more likely to match what the standard library produces.

Depending on the system and configuration, results may not match
exactly, at least with DNSPython. There are still some rare cases
where the system resolver can raise ``herror`` but DNSPython will
raise ``gaierror`` or vice versa. There doesn't seem to be a
deterministic way to account for this. On PyPy, ``getnameinfo`` can
produce results when CPython raises ``socket.error``, and gevent's
DNSPython resolver also raises ``socket.error``.

In addition, several other small discrepancies were addressed,
including handling of localhost and broadcast host names.

.. note:: This has been tested on Linux (CentOS and Ubuntu), macOS,
          and Windows. It hasn't been tested on other platforms, so
          results are unknown for them. The c-ares support, in
          particular, is using some additional socket functions and
          defines. Please let the maintainers know if this introduces
          issues.
